### PR TITLE
Make solrupdater handle works without titles

### PR DESCRIPTION
**Pull request by cdrini**
This is the most logged error from solr updater. I deleted some of the clearly nonsense ones, but some of them are valid works, so just added a simple checks to reduce the noise in the logs.

Extracted from #4337 

### Testing
- Unit tests should be enough.

### CODE REVIEW

**Comment by hornc** [comment]
@cdrini do you have some examples of works without titles? Works without titles sound like a problem for the data model, and with data entry. I can't think of a case where that could be valid.

**Comment by cdrini** [comment]
Here are some examples; not sure if this list is exhaustive. Some of these are easy to cleanup since they have 1+ editions with a title, but some also have an edition without a title 😬 ; those will need manual intervention.

- https://openlibrary.org/works/OL2133301W
- https://openlibrary.org/works/OL21371902W
- https://openlibrary.org/works/OL21371934W
- https://openlibrary.org/works/OL21371945W
- https://openlibrary.org/works/OL21371899W
- https://openlibrary.org/works/OL21371901W
- https://openlibrary.org/works/OL21371903W
- https://openlibrary.org/works/OL21371900W
- https://openlibrary.org/works/OL21371904W

<details>
<summary>More</summary>

- https://openlibrary.org/works/OL21723722W
- https://openlibrary.org/works/OL2182143W
- https://openlibrary.org/works/OL2300653W
- https://openlibrary.org/works/OL2421871W
- https://openlibrary.org/works/OL5042160W
- https://openlibrary.org/works/OL5376901W
- https://openlibrary.org/works/OL5376902W
- https://openlibrary.org/works/OL6651712W
- https://openlibrary.org/works/OL8146996W
- https://openlibrary.org/works/OL862048W
- https://openlibrary.org/works/OL9727763W
- https://openlibrary.org/works/OL12333938W
- https://openlibrary.org/works/OL12850561W
- https://openlibrary.org/works/OL13230236W
- https://openlibrary.org/works/OL15122312W
- https://openlibrary.org/works/OL15165266W
- https://openlibrary.org/works/OL15165836W
- https://openlibrary.org/works/OL15166616W
- https://openlibrary.org/works/OL15166375W
- https://openlibrary.org/works/OL15166492W
- https://openlibrary.org/works/OL15166438W
- https://openlibrary.org/works/OL15167769W
- https://openlibrary.org/works/OL15167264W
- https://openlibrary.org/works/OL15167947W
- https://openlibrary.org/works/OL15168199W
- https://openlibrary.org/works/OL15168583W
- https://openlibrary.org/works/OL15168044W
- https://openlibrary.org/works/OL15168062W
- https://openlibrary.org/works/OL15168233W
- https://openlibrary.org/works/OL15168582W
- https://openlibrary.org/works/OL15168690W
- https://openlibrary.org/works/OL15297711W
- https://openlibrary.org/works/OL15371973W
- https://openlibrary.org/works/OL15376378W
- https://openlibrary.org/works/OL15392318W
- https://openlibrary.org/works/OL15394323W
- https://openlibrary.org/works/OL15399398W
- https://openlibrary.org/works/OL15399039W
- https://openlibrary.org/works/OL15659191W
- https://openlibrary.org/works/OL15670635W
- https://openlibrary.org/works/OL15676125W
- https://openlibrary.org/works/OL15676037W
- https://openlibrary.org/works/OL15749122W
- https://openlibrary.org/works/OL16080289W
- https://openlibrary.org/works/OL16223986W
- https://openlibrary.org/works/OL16799126W
- https://openlibrary.org/works/OL16799127W
- https://openlibrary.org/works/OL16803850W
- https://openlibrary.org/works/OL16806101W
- https://openlibrary.org/works/OL16806562W
- https://openlibrary.org/works/OL16806456W
- https://openlibrary.org/works/OL16806768W
- https://openlibrary.org/works/OL16806651W
- https://openlibrary.org/works/OL16808660W
- https://openlibrary.org/works/OL16811817W
- https://openlibrary.org/works/OL16830951W
- https://openlibrary.org/works/OL16830942W
- https://openlibrary.org/works/OL16830964W
- https://openlibrary.org/works/OL16830929W
- https://openlibrary.org/works/OL16830959W
- https://openlibrary.org/works/OL16830580W
- https://openlibrary.org/works/OL16830936W
- https://openlibrary.org/works/OL16830932W
- https://openlibrary.org/works/OL16830945W
- https://openlibrary.org/works/OL16830955W
- https://openlibrary.org/works/OL16830954W
- https://openlibrary.org/works/OL16834249W
- https://openlibrary.org/works/OL16850327W
- https://openlibrary.org/works/OL16896540W
- https://openlibrary.org/works/OL10652818W
- https://openlibrary.org/works/OL17877852W
- https://openlibrary.org/works/OL18191882W
- https://openlibrary.org/works/OL19089370W
- https://openlibrary.org/works/OL19336378W
- https://openlibrary.org/works/OL19792039W
- https://openlibrary.org/works/OL20087875W
- https://openlibrary.org/works/OL20123775W
- https://openlibrary.org/works/OL20476129W

</details>

File: openlibrary/solr/update_work.py [hunk 1]
```diff
@@ -731,8 +731,13 @@ def build_data2(w, editions, authors, ia, duplicates):
...
-    title = w.get('title', None)
-    if not title:
+    # Some works are missing a title, but have titles on their editions
+    w['title'] = next(itertools.chain(
+        (book['title'] for book in itertools.chain([w], editions) if book.get('title')),
+        [None]
+    ))
+    if not w['title']:
+        logger.error('Work missing title %s' % w['key'])
         return
```
Comment (cdrini) [inline openlibrary/solr/update_work.py:741]:
```suggestion
    w['title'] = next(itertools.chain(
        (book['title'] for book in itertools.chain([w], editions) if book.get('title')),
        ['__None__']
    ))
    if w['title'] == '__None__':
        logger.warn('Work missing title %s' % w['key'])
```

**Comment by mekarpeles** [comment]
lgtm

**Pull request merged**